src/www/app/server/access_key.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Config, makeConfig, SHADOWSOCKS_URI, SIP002_URI} from 'ShadowsocksConfig';
+import {ServerUrlInvalid} from '../../model/errors';
+import {ShadowsocksConfig} from '../config';
+
+export class OutlineServerAccessKey implements ShadowsocksConfig {
+  _rawConfig: Config;
+
+  host?: string;
+  port?: number;
+  method?: string;
+  password?: string;
+  name?: string;
+
+  isOutlineServer = false;
+
+  static fromConfig({name, ...config}: ShadowsocksConfig) {
```

Comment:
We could also flip this? The main constructor takes a config object, then we have `fromShadowsocksUri`

Ah, you know what, I think the main constructor should just take either or, no?